core/server/worker/src/main/java/alluxio/worker/block/UfsIOManager.java

```diff
@@ -0,0 +1,298 @@
+        Thread.currentThread().interrupt();
+      }
+      if (task != null) {
+        int quota = mThroughputQuota.getOrDefault(task.getUser(), -1);
+        double rate = getUsedThroughput(task.mUser);
+        if (quota == -1 || quota * Constants.MB > rate) {
+          try {
+            mUfsIoExecutor.submit(task);
+          } catch (RejectedExecutionException e) {
+            // should not reach here since we use unbounded queue in thread pool and use a bounded
+            // blocking queue upfront to control number of read tasks
+            mReadQueue.add(task);
+          }
+        } else { // resubmit to queue
+          mReadQueue.add(task);
+        }
+      }
+    }
+  }
+
+  @VisibleForTesting
+  public double getUsedThroughput(String user) {
+    Meter ufsBytesReadThroughput = mUfsBytesReadThroughputMetrics.computeIfAbsent(
+        new BytesReadMetricKey(mUfsClient.getUfsMountPointUri(), user),
+        key -> key.getUser() == null
+            ? MetricsSystem.meterWithTags(MetricKey.WORKER_BYTES_READ_UFS.getName(),
+            MetricKey.WORKER_BYTES_READ_UFS.isClusterAggregated(), MetricInfo.TAG_UFS,
+            MetricsSystem.escape(key.getUri()))
+            : MetricsSystem.meterWithTags(MetricKey.WORKER_BYTES_READ_UFS.getName(),
+            MetricKey.WORKER_BYTES_READ_UFS.isClusterAggregated(), MetricInfo.TAG_UFS,
+            MetricsSystem.escape(key.getUri()), MetricInfo.TAG_USER, key.getUser()));
+    return ufsBytesReadThroughput.getOneMinuteRate();
+  }
+
+  /**
+   * read from ufs.
+   * 
+   * @param blockId block id
+   * @param blockSize block size
+   * @param ufsPath ufs path
+   * @param offset offset to read
+   * @param length length to read
+   * @param positionShort is position short or not
+   * @param user user tag
+   * @return content read
+   * @throws ResourceExhaustedException any exception from ufs
+   */
+  public CompletableFuture<byte[]> read(long blockId, long blockSize, String ufsPath, long offset,
+      long length, boolean positionShort, String user) throws ResourceExhaustedException {
+    CompletableFuture<byte[]> future = new CompletableFuture<>();
+    if (mReadQueue.size() >= READ_CAPACITY) {
+      throw new ResourceExhaustedException(format("UFS read at capacity, try later or increase %s",
+          PropertyKey.UNDERFS_IO_READ_QUEUE_CAPACITY.getName()));
+    }
+    mReadQueue.add(
+        new ReadTask(blockId, blockSize, ufsPath, offset, length, positionShort, user, future));
+    return future;
+  }
+
+  private class ReadTask implements Runnable {
+    private final long mOffset;
+    private final long mLength;
+    private final boolean mIsPositionShort;
+    private final CompletableFuture<byte[]> mFuture;
+    private final long mBlockId;
+    private final String mUfsPath;
+    private final String mUser;
+    private final long mBlockSize;
+
+    private ReadTask(long blockId, long blockSize, String ufsPath, long offset, long length,
```

Comment:
This is more of a style issue. Personally I'd prefer to have a clearer API. `ReadTask` needs ufsPath, start/offset, bytesToRead, fileId, positionShort, metrics, so construct those and initialize the task with better capsulation would make the code easier to read. Otherwise the caller of this constructor is just dumping information.